### ISSUE TITLE
Do contextual logging for scheduler

### ISSUE THREAD

**Issue by damemi** [kubernetes/kubernetes#91633]
The kube-scheduler is currently difficult to debug through logs. Logging is inconsistent and sparse, and critically helpful information is in some spots logged at too high of a level to be practically accessible (such as [node scores](https://github.com/kubernetes/kubernetes/blob/b618411f1edb98afe807866fb3a607356d72ba24/pkg/scheduler/core/generic_scheduler.go#L703), which are `V(10)`. At this level logs are flooded with low-level information too quickly to even grep).

**What would you like to be added**:
Standardize scheduler logging to useful, documented levels to improve debuggability and diagnoses of scheduling anomalies. This information should be accessible at reasonable levels so as not to be flooded by irrelevant low-level process and network logs (ie, `v=2,3,4..`) and consistent across the scheduler.

Specifically, information that is necessary includes, but is not limited to:
* Filter plugins
  * Per plugin/per node
* Preemption attempts (/postfilter plugins)
* Score plugins
  * Normalized results
    * Per plugin/per node
* Nominated node result
(sub-lists to indicate potential logging hierarchies. these are just some examples)

In addition, these logs should be able to be easily traced through a scheduling cycle. That is, all logs related to a specific scheduling cycle should have a unique identifier so that relevant information can be efficiently gathered with common search tools like grep.

**Why is this needed**:
The most common issues with the scheduler are questions related to the scheduling cycle ("Why did/didn't my pod get scheduled here/there?"). While the scheduler does currently report `FailedScheduling` events, these can lack information (https://github.com/kubernetes/kubernetes/issues/91340, https://github.com/kubernetes/kubernetes/issues/91601) and do not provide any insight to successful scheduling that does not land on the expected node (eg, scoring interference).

There is demand to be able to easily debug the steps of the scheduling cycle, and the lack of that ability frequently confuses users. It would be very helpful to be able to say, "If you set `v=2`, you'll be able to see each plugin's results" for example.

/sig scheduling

**Comment by damemi** [comment]
cc @ahg-g @Huang-Wei @alculquicondor @ingvagabund @ravisantoshgudimetla wdyt? To make sure our logging stays consistent I think it would be helpful to have some kind of design to point to

**Comment by ahg-g** [comment]
+1

I like the idea of associating the logs with an identifier. We should look at structured logging and see if what you are planning to do aligns with that proposal: https://github.com/kubernetes/enhancements/issues/1602

**Comment by alculquicondor** [comment]
+1 on structured logging.

**Comment by ingvagabund** [comment]
+1
I wish we could have two logging streams, one of them dedicated to just scheduling decisions.

**Comment by Huang-Wei** [comment]
+1.

In addition to structured logging (to have consistent identifiers), we should have consistent mechanics to define each log level's semantics. No need to have a KEP, but would be nice to keep it documented in https://github.com/kubernetes/community/tree/master/contributors/devel/sig-scheduling.

**Comment by serathius** [comment]
Definitely it would be good to have cooperation between SIG Instrumentation and Scheduling to collect set requirements for structured logging.
Defining log semantics (aka log schema) is definitely on our plate (was part of original proposal, but was postponed to focus on migration). 
As for separating log streams for scheduling decisions, it would be interesting to use more https://github.com/go-logr/logr (generic logging interface that we are using underneath klog) features like named loggers

Adding logr authors to discussion.
/cc @thockin  @DirectXMan12

**Comment by soltysh** [comment]
> /assign

@jherrera123 I'd suggest syncing with @damemi on this, since he initiated this bug to start working on it, as far as I can tell.

**Comment by damemi** [comment]
Thanks for the discussion so far everyone. I wasn't totally aware of the structured logging KEP, but I think this does align well with it. I see that there are even sections of the scheduler that have already been identified as good spots to add structured logging.

Like @Huang-Wei said, in addition to logging identifiers it's important that we document consistent log levels so users can know exactly what they're getting at each level. I think this goes beyond the existing structured logging KEP (and is scheduler-specific), so this issue isn't quite a duplicate of the KEP.

and @jherrera123 I haven't started work on this yet (still getting feedback here) but I won't turn down any offers to help :)

**Comment by jherrera123** [comment]
@damemi once we land on a design, I'd be happy to help. It looks like there's a lot to be done in the logging area.

**Comment by damemi** [comment]
Sorry all, haven't forgotten about this. Read up some more on the structured logging and it doesn't look too bad to implement. I think the big thing (at least for the scheduler) would be something like the named loggers @serathius mentioned in https://github.com/kubernetes/kubernetes/issues/91633#issuecomment-638137616. Though, I think we could get sufficient context with just a pod key in the structured logs, at least for now.

**Comment by alculquicondor** [comment]
@damemi could you make a list of what's needed to have kube-scheduler in a desirable state with regards to structured logging?

**Comment by damemi** [comment]
> @damemi could you make a list of what's needed to have kube-scheduler in a desirable state with regard to structured logging?

Yeah, I'll work on putting together what we talked about here. The main improvement would be logging a unique key for each pod through its scheduling cycle like mentioned above

**Comment by pohly** [comment]
The solution with structured logging would be to also do contextual logging:
```
logger := logger.WithValues("pod", klog.KObj(pod))
```

Then all `klog.InfoS` call needs to be replaced with `logger.Info` and the pod key/value would always be included in each log message, even in code which isn't even aware that it is being called for the pod. It's also possible to attach a name like "scheduler" or "scheduler/predicate-XYZ".

The problem in the Kubernetes code base is that there's no accepted mechanism for propagating a logger instance from the caller to the callee. This is a long-term goal which still needs a KEP.

**Comment by pohly** [comment]
I have published a PR with the KEP and intend to discuss it at this week's SIG Instrumentation meeting: https://github.com/kubernetes/enhancements/pull/3078

The first step will be to decide whether the SIG wants to tackle this problem.

@damemi I've used [your comment above](https://github.com/kubernetes/kubernetes/issues/91633#issuecomment-675074671) as basis for one of the use cases in the KEP.

**Comment by alculquicondor** [comment]
cc @serathius 

We are onboard. Several contributors helped migrating to structured logging and I expect they would like to contribute further.

**Comment by damemi** [comment]
+1 I think the use case is valid for your proposal @pohly

**Comment by weberc2-tempus** [comment]
In the meanwhile, is there a way to get scoring lines using the `--v` and `--vmodule` flags? Specifically, I thought `kubectl logs ... -vmodule=generic_scheduler.go=10` would activate [this log line](https://github.com/kubernetes/kubernetes/blob/release-1.19/pkg/scheduler/core/generic_scheduler.go#L524-L528) but it does not (at least not for kube-scheduler=v1.19.7):

```go
if klog.V(10).Enabled() {
	for plugin, nodeScoreList := range scoresMap {
		klog.Infof("Plugin %s scores on %v/%v => %v", plugin, pod.Namespace, pod.Name, nodeScoreList)
	}
}
```

**Comment by alculquicondor** [comment]
I haven't used `vmodule`, but it seems that you are not supposed to put `.go` https://pkg.go.dev/k8s.io/klog#pkg-overview

**Comment by pohly** [comment]
My goal for this issue is to get https://github.com/kubernetes/enhancements/pull/3078 and the core infrastructure merged. I already have a prototype conversion of some parts of kube-scheduler (see KEP), but will probably need help from folks in SIG Scheduling to finish that part.

**Comment by alculquicondor** [comment]
You'll have the full support of reviewers. I'm confident that we can find contributors to help you with some changes.

**Comment by alculquicondor** [comment]
Whenever the infrastructure is ready, we can create some fine-grained issues for contributors to pick up.

**Comment by weberc2-tempus** [comment]
Cc @alculquicondor

> I haven't used vmodule, but it seems that you are not supposed to put .go

I tried passing `--v=4 --vmodule=generic_scheduler=10` directly into the pod spec (as opposed to via kubectl) and that doesn't seem to work either. I've also tried `--v=10` and nothing is ever printed from generic_scheduler.go.

**Comment by alculquicondor** [comment]
Please don't ask for further support here. You can use the slack or other means.

**Comment by Huang-Wei** [comment]
I'm thinking if it's viable to dump detailed pod's scheduling logs dynamically, on pod's basis. Here is a draft idea: dump the logs (e.g., [score logs](https://github.com/kubernetes/kubernetes/blob/4d08582d1fa21e1f5887e73380001ac827371553/pkg/scheduler/scheduler.go#L1126-L1134)) when a pod carries particular labels/annotations: like "alpha.k8s.io/scheduler-log-level: v10".

The takeaway is in this way, the k8s users can enable logging upon suspicious scheduling behavior on demand, without requesting cluster SRE to uplift the global logging level. However, an argument is this may be abused to be malicious, esp. in a multi-tenant env like public k8s managed service. WDYT?

**Comment by serathius** [comment]
> The takeaway is in this way, the k8s users can enable logging upon suspicious scheduling behavior on demand, without requesting cluster SRE to uplift the global logging level. However, an argument is this may be abused to be malicious, esp. in a multi-tenant env like public k8s managed service. WDYT?

I don't think that k8s user should normally have access to logs. They should still go through cluster SRE to get the scheduler logs. Log verbosity is set in command line flags, as so it normally requires user to be a cluster admin. If we want to implement overriding log verbosity per pod, it should should require a cluster admin level permissions.

**Comment by Huang-Wei** [comment]
> I don't think that k8s user should normally have access to logs.

It's correct. I was proposing that k8s users toggle enabling logs for suspicious pods, and then the SRE to grab the logs.

> it should should require a cluster admin level permissions.

We have ways to work it around, like abstracting verbose log as a GVR (under pod's /subresource for example), and make RBAC protected. This is the next step we can discuss after getting a consensus on the "dynamic logging" idea.

**Comment by alculquicondor** [comment]
if it's a subresource, it should probably be a field, instead of an annotation. But also it means you cannot set it during Pod creation. So you could potentially be missing the logs for the first scheduling attempt anyways.

**Comment by damemi** [comment]
Even from the SRE's standpoint, being able to have a higher log level for specific pods would make filtering through the logs much easier. It would also reduce the noise that you get from actually setting `--v=10`.

**Comment by alculquicondor** [comment]
@liggitt do you have any guidance from a sig-auth perspective forhttps://github.com/kubernetes/kubernetes/issues/91633#issuecomment-1071741610 ?

**Comment by liggitt** [comment]
pod authors should not have control over scheduler log verbosity... that can expose the scheduler to DOS attacks

**Comment by alculquicondor** [comment]
What if we have a quota scope for it?

**Comment by liggitt** [comment]
that seems like scheduler-level configuration, which I would not expect to be driven by the pod API

**Comment by alculquicondor** [comment]
Yeah, but kube-scheduler configuration is global (affects all pods) and requires restart.

Restarting would clear the cache, which would potentially make the bug we are trying to diagnose disappear.

Here is another alternative: We allow the configuration of a file where SREs can put pod names (or prefixes). We probably need to watch for changes in the file.

**Comment by ravisantoshgudimetla** [comment]
Trying to understand here. What's wrong with having a virtual apigroup to support RBAC or new field in the pod spec which is RBAC protected. Quota for cases which can impact storage(via increased scheduler logging in this case) is hard to count/measure.

**Comment by Huang-Wei** [comment]
A safe solution to toggle verbosed pod logging would be super helpful to scheduler, and also to the whole k8s ecosystem.

> Trying to understand here. What's wrong with having a virtual apigroup to support RBAC or new field in the pod spec which is RBAC protected.

+1, that's what i meant as GVR (not having a physical storage backing it).

**Comment by ravisantoshgudimetla** [comment]
Any update on this? I think being behind RBAC would be super helpful from developer and admin standpoint.

**Comment by liggitt** [comment]
I don't think it's appropriate to put control over scheduler config/operation on pod-level API like this.

**Comment by damemi** [comment]
Not trying to be pedantic, but if you consider things like affinity, nodeselector, podtopology etc there are already scheduler options on the pod level. These are meant to control individual pod-specific behavior (eg we don't just require one top-down nodeaffinity setting)

I see the security risk but if this is behind good RBAC then are there other differences between settings like that and something like this?

**Comment by smarterclayton** [comment]
This sounds like “debugging info per pod for scheduler / cluster admins”, not “debugging info for users who don’t understand why pods weren’t placed”, especially when we’re talking about the log output of the scheduler.  The latter is generally a much higher bar and doesn’t map to 100% of cluster types (those where end users are heavily isolated from scheduler logs).

At some level, for the former why doesn’t the scheduler expose the equivalent simulation of a schedule via an api on the scheduler itself?  Ie “hey scheduler, tell me why X was scheduled” with output coming from the scheduler process as some form of structured or unstructured output?  That’s better than logs for running pods, and would control the security of the output (you could simply duplicate the effective process log output to start) without causing an attack vector for pod authors?

**Comment by ravisantoshgudimetla** [comment]
@liggitt - I agree with your point however I think that this is a problem of UX vs security(possibility of DOS). To me the problem here is age old and we have never thought of scheduling as a security boundary. Just to add to what @damemi suggested, some user can set nodeName or nodeAffinity, podAffinity, tolerations etc on the pod and can still DOS the node. 

Thinking out loud here, if you're saying that having RBAC on the scheduling config within pod spec isn't appropriate would adding a new field to pod spec(which encompasses all scheduling constraints and have RBAC on the various resources within it) be a viable solution? This approach can set precedent on who can set scheduling constraints on the pod so that we don't overwhelm the node in future. I am asking as I want to know the alternatives, the only solution we have to limit is via quota and in this case it may be hard to do limit via quota.


>  “debugging info per pod for scheduler / cluster admins”, not “debugging info for users who don’t understand why pods weren’t placed”,

@smarterclayton - The issue is not only specific to cluster-admin, the end-users are given minimal information from which they cannot make out which scheduling constraints can be relaxed/updated or why a particular node is chosen. As end-user, I care about such information as well because I want to know why my pod is not getting scheduled or why a particular node is chosen. 

> At some level, for the former why doesn’t the scheduler expose the equivalent simulation of a schedule via an api on the scheduler itself? Ie “hey scheduler, tell me why X was scheduled” with output coming from the scheduler process as some form of structured or unstructured output? That’s better than logs for running pods, and would control the security of the output (you could simply duplicate the effective process log output to start) without causing an attack vector for pod authors?

We can have simulator but having real time data for a pod would be beneficial from both cluster-admin and end-user pov. We don't need logs of the long running pods, we just need the logs of the scheduler with increased verbosity during that time.

**Comment by liggitt** [comment]
Fields on the pod to control where they are scheduled makes sense to me. Fields on the pod to control debug settings on the scheduler doing the scheduling does not make sense to me.

**Comment by pohly** [comment]
Let me bring up a different approach. Instead of extending the Pod spec, a new, separate `SchedulePod` object could be created that references a Pod. RBAC permissions for that new type then can be chosen as intended for the cluster, for example only allow an admin to create such objects.

When kube-scheduler starts a scheduling attempt for a Pod, it would need to check for a matching `SchedulePod` object. It then can increase the verbosity and emit log entries to the normal log. It can also go one step further and use a custom logger that captures the log output for that one Pod in addition to sending it to the global log. Then the output can be stored separately or get added to the `SchedulePod` object for the Pod. That would solve the problem of "how can admins (or users) access the log output".

**Comment by knelasevero** [comment]
So, based on @pohly's [PR](https://github.com/kubernetes/kubernetes/pull/110833) I wanted to start by getting some values out from the options that we have there in those commits. Took me a while since these benchmark tests make my laptop useless while they are running and they take a while (and I have to do some other stuff :sweat_smile:).

- First commit there is bringing support for contextual logging: https://github.com/kubernetes/kubernetes/pull/110833/commits/e0b9d79758b1fe7b2453c275a783cacd94ba7a15

Here is the pprof cpu profiling output: https://github.com/knelasevero/scheduler-pprof-results/blob/main/cpu_prof/cpu-profile-e0b9d79758b1fe7b2453c275a783cacd94ba7a15.out

Here is the generated graph (click raw, since github wont show the pre-visualization): 
https://github.com/knelasevero/scheduler-pprof-results/blob/main/cpu_prof/e0b9d79758b1fe7b2453c275a783cacd94ba7a15.svg

- Second commit is adding json support in performance tests:
https://github.com/kubernetes/kubernetes/pull/110833/commits/408a28e76769b05636fa63f1e1fdef13cdc913be

Here is the pprof cpu profiling output: 
https://github.com/knelasevero/scheduler-pprof-results/blob/main/cpu_prof/cpu-profile-408a28e76769b05636fa63f1e1fdef13cdc913be.out

Here is the generated graph:
https://github.com/knelasevero/scheduler-pprof-results/blob/main/cpu_prof/408a28e76769b05636fa63f1e1fdef13cdc913be.svg

- Third commit is adding WithName and WithValues, and with that, a prefix with operation and which component is generating a log entry:
https://github.com/kubernetes/kubernetes/pull/110833/commits/01b8046c58e288a0a9d171235fb37d4abe0acc3e

Here is the pprof cpu profiling output: 
https://github.com/knelasevero/scheduler-pprof-results/blob/main/cpu_prof/cpu-profile-01b8046c58e288a0a9d171235fb37d4abe0acc3e.out

Here is the generated graph:
https://github.com/knelasevero/scheduler-pprof-results/blob/main/cpu_prof/01b8046c58e288a0a9d171235fb37d4abe0acc3e.svg

- Forth commit is first attempt at reducing overhead by only only modifying loggers at v>3:
https://github.com/kubernetes/kubernetes/pull/110833/commits/32596ed3ef9f75fec5573c7a0c1dab5949a7d16e

 Here is the pprof cpu profiling output: 
 https://github.com/knelasevero/scheduler-pprof-results/blob/main/cpu_prof/cpu-profile-32596ed3ef9f75fec5573c7a0c1dab5949a7d16e.out
 
 Here is the generated graph:
 https://github.com/knelasevero/scheduler-pprof-results/blob/main/cpu_prof/32596ed3ef9f75fec5573c7a0c1dab5949a7d16e.svg
 
 - Fifth commit is seccond attempt at reducing overhead by only including pod and node, with no WithName:
 https://github.com/kubernetes/kubernetes/pull/110833/commits/f7ac1e31be7fe27730b6a7dace2fdb35490e5f5c
 
 Here is the pprof cpu profiling output: 
 https://github.com/knelasevero/scheduler-pprof-results/blob/main/cpu_prof/cpu-profile-f7ac1e31be7fe27730b6a7dace2fdb35490e5f5c.out
  
 Here is the generated graph:
 https://github.com/knelasevero/scheduler-pprof-results/blob/main/cpu_prof/f7ac1e31be7fe27730b6a7dace2fdb35490e5f5c.svg
 
 I can generate some other benchmarks, which do you think would be relevant?
 
 @pohly was this the info that you expected to see?
 
 Also, open question here, how should we go about deciding the right trade-off between information in log entries versus runtime performance?

**Comment by alculquicondor** [comment]
Can you summarize the findings in a sheet or something equivalent?

**Comment by knelasevero** [comment]
Sure, on it

**Comment by pohly** [comment]
> your second commit reads as `scheduler_perf: support JSON` but I got BenchmarkPerfScheduling* files also for the first commit. Should I be getting another output somewhere?

That commits adds the same `--logging-format=[json|text]` parameter that other Kubernetes binaries have. Without that parameter it always writes logs as text.

Probably has little influence on the instrumentation overhead.

**Comment by knelasevero** [comment]
Hey, I reviewed the tests, and re-ran them in a machine where I could be sure that it would not have interference during the tests. New results are already in that doc: https://docs.google.com/spreadsheets/d/1lqUDnx6WRGB5Uu1-Xgaa_sG1bQsINA_SZ2fkay38gWs/edit#gid=0

@pohly  I also creatred another PR pointing to what I think is the commit that we should pursue :sweat_smile:. Also rebased and fixed conflicts. Please let me know if I should have rebased differently there. https://github.com/kubernetes/kubernetes/pull/111155

**Comment by alculquicondor** [comment]
Sorry, I have been to busy with other features in the SIG. I quickly checked the spreadsheet (thank you for humoring me).

Can you clarify at which log level did you run the experiments?

If you did it at v=2, I suppose the results for `only v>3` is kind of the baseline. If so, it kind of looks like with json the performance is quite similar.
But if we start using WithName, WithValues, the performance is significantly impacted (more than 50% overhead in some cases).
Did I interpret the results right?

**Comment by knelasevero** [comment]
> Sorry, I have been to busy with other features in the SIG

No worries

>  I quickly checked the spreadsheet (thank you for humoring me)

:+1: 

> Can you clarify at which log level did you run the experiments?

Since I wanted the cpu profile at every run that I ran, I was running this command: 

```
make test-integration WHAT=./test/integration/scheduler_perf KUBE_TIMEOUT="-timeout=3600s" ETCD_LOGLEVEL=warn KUBE_TEST_VMODULE="''" KUBE_TEST_ARGS="-alsologtostderr=false -logtostderr=false -run=^$$ -benchtime=1ns -bench=BenchmarkPerfScheduling -cpuprofile ~/cpu-profile.out"
```

So I was not setting the loglevel explicitly, so I was using default loglevel. But yeah, I think first commits + v>3 are kind of the baseline for the reason that you stated.

> But if we start using WithName, WithValues, the performance is significantly impacted (more than 50% overhead in some cases).
Did I interpret the results right?

It is `more than 50% overhead in some cases` but just a few, right? Most rows have a similar performance there or something like 20 or 30% impact. I am not sure if there are some of those tests that are more relevant and should be considered with more weight. Is that the case?

**Comment by alculquicondor** [comment]
Probably `BenchmarkPerfScheduling/SchedulingBasic/5000Nodes` can be used as the primary reference. Can you run this benchmark without any changes to see the actual baseline?

**Comment by knelasevero** [comment]
@alculquicondor I ran another perf test from current master branch. added columns between `01b8046c58e - WithName WithValues` and `32596ed3ef9 - only v>3` in the spreadsheet. It is tagged `master branch` there.

Some results are still exploding, I guess when something hangs. How were perf tests ran before in the past to grantee sensible results? I guess using a standard empty cloud VM would be better than running them on my machine if that is required.

In any case, most results still seem similar with v>3 commit. Do we think using WithName, WithValues impacts performance too much?

**Comment by alculquicondor** [comment]
FTR, the most important metric is `scheduler_scheduling_attempt_duration_seconds`, which doesn't seem to be affected by too much. I think we can proceed with caution and be ready to revert if our CI performance dashboard shows a significant regression.

http://perf-dash.k8s.io/#/?jobname=kubemark-5000Nodes&metriccategoryname=Scheduler&metricname=SchedulingLatency&Operation=scheduling&TestName=load

@ahg-g @Huang-Wei in case you want to add more thoughts.

**Comment by Huang-Wei** [comment]
The only concern is whether we generate enough volume of contextual logging calls. Can we get some data on that? like grep the contextual logging string in the benchmark test log.

**Comment by knelasevero** [comment]
Watched the sig meeting. I will try to figure out a way to report how much contextual logging calls (or more specifically how much getting-logger-from-ctx happens) in the tests. Or, as you @Huang-Wei mentioned, I will try to add a scenario where we are a bit sure that we are hitting it sufficiently to compare performance degradation with and without the actual code changes.

**Comment by knelasevero** [comment]
Hey, so @Huang-Wei I have run perf tests only for PreemptionBasic which I believe will basically always trigger getting a logger from context and logging something at basically any loglevel (as you mentioned). I ran a few times both with code from master branch and also from the branch using ctx logger with WithName and with values. Results are here: https://gist.github.com/knelasevero/79b7fa6f57c4b3cc863cd9fb8761269b

**Comment by Huang-Wei** [comment]
> Seems to oscillate enough for it to be even faster sometimes with WithValues/WithNames on my machine

:) That may be related to the actual system loading on the testing machine. The scalability test on perf-dash.k8s.io should be much more reliable.

And forgot to mention the general Golang benchmark testing, we can write tests for the core WithXYZ() logging functions. That should tell the perf diff in a function's level. (ignore it if it's already implemented)

BTW: where are we in KEP https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging?

**Comment by pohly** [comment]
> And forgot to mention the general Golang benchmark testing, we can write tests for the core WithXYZ() logging functions. That should tell the perf diff in a function's level. (ignore it if it's already implemented)

What kind of conclusion would one draw from measuring individual functions? How slow is "too slow"?

> where are we in KEP https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging?

The infrastructure is complete, but still marked as alpha and off by default. I've been using it in various projects without any problems, so I think it is stable. What's missing is some real usage and feedback by others. kube-scheduler is leading here.

**Comment by Huang-Wei** [comment]
> What kind of conclusion would one draw from measuring individual functions? How slow is "too slow"?

I'd say a 10% perf drop on individual function's benchmark testing is acceptable - given klog calls don't contribute much to the overall scheduling time.

**Comment by pohly** [comment]
So you meant individual functions in the scheduler, not the WithName/Values functions - sorry, I misunderstood.

**Comment by Huang-Wei** [comment]
yup, choosing a function that covers WithName/Value functions call and Logger's pass-in would be nice.

**Comment by knelasevero** [comment]
> And forgot to mention the general Golang benchmark testing, we can write tests for the core WithXYZ() logging functions. That should tell the perf diff in a function's level. (ignore it if it's already implemented)

> And forgot to mention the general Golang benchmark testing, we can write tests for the core WithXYZ() logging functions. That should tell the perf diff in a function's level. (ignore it if it's already implemented)

I am not familiar with writing perf test with how things are wired here in this repo. You mean benchmark tests of functions instead of benchmark tests that are integration tests (like they are right now, there)?

**Comment by Huang-Wei** [comment]
> You mean benchmark tests of functions instead of benchmark tests that are integration tests (like they are right now, there)?

Not quite "replacing" the benchmark integration test, it's complementary :)

Here are some examples in scheduler's codebase that benchmark a particular function:

```
⇒  ag benchmark pkg/scheduler
pkg/scheduler/framework/plugins/selectorspread/selector_spread_perf_test.go
52:func BenchmarkTestSelectorSpreadPriority(b *testing.B) {

pkg/scheduler/framework/plugins/volumebinding/binder.go
938:	// TODO (for beta): benchmark this and potentially introduce some kind of lookup structure (https://github.com/kubernetes/enhancements/issues/1698#issuecomment-654356718).

pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go
1272:func BenchmarkTestPodTopologySpreadScore(b *testing.B) {
1371:func BenchmarkTestDefaultEvenPodsSpreadPriority(b *testing.B) {

pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go
2206:func BenchmarkFilter(b *testing.B) {

pkg/scheduler/internal/cache/cache_test.go
1857:func BenchmarkUpdate1kNodes30kPods(b *testing.B) {
1866:func BenchmarkExpirePods(b *testing.B) {
1875:			benchmarkExpire(b, podNum)
1880:func benchmarkExpire(b *testing.B, podNum int) {

pkg/scheduler/internal/queue/scheduling_queue_test.go
449:func BenchmarkMoveAllToActiveOrBackoffQueue(b *testing.B) {
```

**Comment by knelasevero** [comment]
I pushed a simple benchmark test coming from master:https://github.com/knelasevero/kubernetes/commit/a66d556ee9bde38ac6103b47a642ec2b34007bce

And I pushed the same benchmark test coming from that [PR](https://github.com/kubernetes/kubernetes/pull/111155):
https://github.com/kubernetes/kubernetes/pull/111155/commits/a44efd83e62c199e7cfc0990744342805dab1416

Running on master:
```
$ git checkout master-with-benchmark
Already on 'master-with-benchmark'
```
```
$ go test -benchmem -run=^$ -bench ^BenchmarkRunPreFilterPlugins$ k8s.io/kubernetes/pkg/scheduler/framework/runtime
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/scheduler/framework/runtime
cpu: Intel(R) Core(TM) i7-10850H CPU @ 2.70GHz
BenchmarkRunPreFilterPlugins/BenchmarkRunPreFilterPlugins-12               65923             16251 ns/op           13635 B/op         57 allocs/op
PASS
ok      k8s.io/kubernetes/pkg/scheduler/framework/runtime       1.352s
```

Running on that PR:

```
$ git checkout log-contextual-scheduler-withnames-withvalues
Switched to branch 'log-contextual-scheduler-withnames-withvalues'
Your branch is up to date with 'knela/log-contextual-scheduler-withnames-withvalues'.
```
```
$ go test -benchmem -run=^$ -bench ^BenchmarkRunPreFilterPlugins$ k8s.io/kubernetes/pkg/scheduler/framework/runtime
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/scheduler/framework/runtime
cpu: Intel(R) Core(TM) i7-10850H CPU @ 2.70GHz
BenchmarkRunPreFilterPlugins/BenchmarkRunPreFilterPlugins-12               62265             24605 ns/op           14153 B/op         80 allocs/op
PASS
ok      k8s.io/kubernetes/pkg/scheduler/framework/runtime       2.974s
```

@Huang-Wei is that what you were looking for?

**Comment by knelasevero** [comment]
Hey @Huang-Wei sorry to poke again 😅 Is this in the direction of what you were looking for?

**Comment by Huang-Wei** [comment]
Yes, that test is exactly I was referring to. Any idea about the performance drop? (from 16251 ns/op -> 24605 ns/op, and I suppose you specified the same logging in both tests)

**Comment by alculquicondor** [comment]
Back to this very old discussion: https://github.com/kubernetes/kubernetes/issues/91633#issuecomment-1071741610

I didn't know other components allowed to change log level at runtime.

https://groups.google.com/a/kubernetes.io/g/dev/c/-AR7YKViP6Y/m/GVKiProsAwAJ?utm_medium=email&utm_source=footer

I don't think we have that in scheduler, do we?

**Comment by knelasevero** [comment]
> Lucas Severo Alves ***@***.***> writes:
> @pohly what do you think here?
> klog text format allows changing the log level at runtime, JSON format (via zapr) doesn't. I don't know whether there is any runtime API (like HTTP) for it. I don't have any strong opinion about whether it should get added.
> […](#)
> -- Best Regards Patrick Ohly Cloud Software Architect

I meant about the performance drop

**Comment by pohly** [comment]
It's impossible to improve logging without paying some price for it. I don't know how important these improvements are and what kind of slowdown would be justified.

**Comment by alculquicondor** [comment]
@knelasevero @pohly we are early in the 1.27 release cycle. I think we should merge what we have and monitor the performance dashboard. If there are issues, we try to optimize in follow up PRs or (worst case) revert.

**Comment by knelasevero** [comment]
@alculquicondor nice

As soon as I am back from vacation I will revise that pr so we can merge. https://github.com/kubernetes/kubernetes/pull/111155

**Comment by Huang-Wei** [comment]
> I don't think we have that in scheduler, do we?

We don't and I don't think other k8s components have either.

We can track this in a separate issue. To have a secure way to change log level at runtime would be pretty useful in the operation's perspective.

**Comment by Huang-Wei** [comment]
@knelasevero if there is any low-hanging fruit we can pick to make the benchmark number better, I will be more confident on the performance. If no luck, I'm fine with an early merge and then count on scalability tests to tell if there is a significant perf downgrade. (btw: it happened a couple of times before)

**Comment by chenk008** [comment]
> > I don't think we have that in scheduler, do we?
> 
> We don't and I don't think other k8s components have either.
> 
> We can track this in a separate issue. To have a secure way to change log level at runtime would be pretty useful in the operation's perspective.

It is very useful to change log level at runtime. When we want to investigate kubelet issue, restarting kubelet will lost a lot information.

**Comment by fromanirh** [comment]
I'd like to echo @chenk008 . Having the ability to dynamically change log levels without restarting components, thus without _having_ to lose context, is a huge quality of life improvement when debugging/troubleshooting live systems.

**Comment by pohly** [comment]
Can we please stop discussing changing log levels in this issue? It's related, but not the same.

**Comment by knelasevero** [comment]
@damemi are you comfortable with making this issue just about contextual logging looking back on what you envisioned for it back in the day? I guess it makes sense to have an issue for contextual logging explicitly since this is what the PR that I have opened resolves, and it should close something. 

But reading the discussion I guess we should split this into even 3 issues {?} The control of log level, the different streams of logs and documenting log levels, and the actual inclusion of contextualized logs? WDYT?

**Comment by damemi** [comment]
@knelasevero yeah, I opened this as a very broad idea of "make the scheduler more transparent" with the intent to promote discussion on how we can do that. I think we have a few ideas now, so narrowing this issue to one of them and tracking others separately is good with me.

**Comment by alculquicondor** [comment]
/help
#115588 needs to be split in a few PRs to facilitate reviews and merge incremental changes.

**Comment by mengjiao-liu** [comment]
> Everything under pkg/scheduler/apis can be its own PR.
>
> The rest is a bit unclear. Maybe it's possible to have a PR for each of the folders in pkg/scheduler/framework/plugins. Followed by a PR that does the rest.

If we start the task, we need to figure out how to split PR.
I am not quite clear about the directory structure of the scheduler, so I will first divide PR from the direction provided by @alculquicondor  and then adjust it according to the actual implementation.

|  Component   | Owner  |
|  ----  | ----  |
|hack/logcheck.conf|@mengjiao-liu |
| pkg/scheduler/apis/  |@mengjiao-liu|
| pkg/scheduler/framework/plugins/defaultbinder/  | @mengjiao-liu |
|pkg/scheduler/framework/plugins/defaultpreemption/ | |
| pkg/scheduler/framework/plugins/examples/stateful/| |
| pkg/scheduler/framework/plugins/interpodaffinity/| |
|pkg/scheduler/framework/plugins/noderesources/ | |
|pkg/scheduler/framework/plugins/nodevolumelimits/ | |
|pkg/scheduler/framework/plugins/podtopologyspread| |
|pkg/scheduler/framework/plugins/volumebinding/ | |
| pkg/scheduler/framework/plugins/volumezone/| |
|pkg/scheduler/internal/cache/ | |
|pkg/scheduler/internal/cache/debugger/ | |
|pkg/scheduler/internal/cache/fake/ | |
|pkg/scheduler/schedule_one.go & pkg/scheduler/scheduler_test.go||
|pkg/scheduler/schedule_one.go & pkg/scheduler/schedule_one_test.go||
|pkg/scheduler/testing/fake_extender.go| |

I'll try to do a part first to see if there will be problems with this division.

**Comment by pohly** [comment]
I would prefer to have one owner for this issue. All of the changes exist, it's just a matter of splitting up the large PR into smaller pieces and then staying on top of concurrent changes to the code.

@mengjiao-liu: if you feel that you can handle this, then I'd be happy to have you work on this and we won't need a second assignee.

**Comment by mengjiao-liu** [comment]
> I would prefer to have one owner for this issue. All of the changes exist, it's just a matter of splitting up the large PR into smaller pieces and then staying on top of concurrent changes to the code.
> 
> @mengjiao-liu: if you feel that you can handle this, then I'd be happy to have you work on this and we won't need a second assignee.

Okay, I can handle this and I'll change the table.

**Comment by alculquicondor** [comment]
what PR should I review first? :)

**Comment by mengjiao-liu** [comment]
> what PR should I review first? :)

I have marked the PRs that have dependencies. PRs that do not write dependencies can be reviewed. You can start with #116842, #116849, or some plugins PRs.

**Comment by mengjiao-liu** [comment]
Work continues in https://github.com/kubernetes/kubernetes/issues/91633#issuecomment-1467645065

/close

**Comment by ziyi-xie** [comment]
Hi @mengjiao-liu ! Is there any other tasks left to be done on this issue? Is there anything in scheduler that hasn't migrated to contextual Logging?

**Comment by mengjiao-liu** [comment]
Sorry, I forgot to update the status of these tasks. The tasks of migrating to contextual logging is complete, see https://github.com/kubernetes/kubernetes/issues/91633#issuecomment-1467645065 for more details.

/close
